### PR TITLE
improve memoryReorder layer to handle conversion between nhwc and nchw

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

improve memoryReorder layer to handle conversion between nhwc and nchw

## How was this patch tested?

unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/2680

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala [hunk 1]
```diff
@@ -26,50 +26,118 @@ class ReorderMemory(inputFormat: MemoryData, outputFormat: MemoryData,
...
+  private var realgradOutput : Array[MemoryData] = null
 
   override private[mkldnn] def initFwdPrimitives(inputs: Array[MemoryData], phase: Phase) = {
     _inputFormats = if (inputFormat == null) inputs else Array(inputFormat)
     require(_inputFormats.length == 1, "Only accept one tensor as input")
 
-    require(_inputFormats(0).shape.product == outputFormat.shape.product,
+    if (outputFormat == null) _outputFormats = _inputFormats
+    require(_inputFormats(0).shape.product == _outputFormats(0).shape.product,
       "input output memory not match")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: memory size not matched, also please output the size accordingly

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala [hunk 2]
```diff
@@ -26,50 +26,118 @@ class ReorderMemory(inputFormat: MemoryData, outputFormat: MemoryData,
...
+    val outputShape = _outputFormats(0).shape
+    val inputLayout = _inputFormats(0).layout
+    val outputLayout = _outputFormats(0).layout
+    realInput = _inputFormats
+    realOutput = _outputFormats
+
+    if (inputLayout != outputLayout) {
+      if (inputLayout == Memory.Format.nhwc) {
+        // remind: if format of input MemoryData is nhwc, its shape should be output shape
+        realInput = Array(HeapData(outputShape, inputLayout))
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: So the input must be a `HeapData` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala [hunk 3]
```diff
@@ -26,50 +26,118 @@ class ReorderMemory(inputFormat: MemoryData, outputFormat: MemoryData,
...
+    val gradOutputShape = _gradOutputFormats(0).shape
+    val gradInputLayout = _gradInputFormats(0).layout
+    val gradOutputLayout = _gradOutputFormats(0).layout
+    realgradInput = _gradInputFormats
+    realgradOutput = _gradOutputFormats
+
+    if (gradInputLayout != gradOutputLayout) {
+      if (gradOutputLayout == Memory.Format.nhwc) {
+        // todo: if format of gradOutput MemoryData is nhwc, its shape should be gradInput shape
+        realgradOutput = Array(HeapData(gradInputShape, gradOutputLayout))
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: Same as above.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: no, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass:
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/291

**Comment by zhangxiaoli73** [comment]
get approval from yanzhang face to face. Do you have any other idea?

**Comment by wzhongyuan** [comment]
ok, please log in github for approval going forward as it may cause confusion to others

**Pull request merged**